File: src/fcc_vm/bin/lexer.ml
type token =
    | PUSH
    | BIND
    | ID of string
    | NUMBER of int 
    | EOF

(* Main lexer function that returns a list of tokens *)
let rec lexe src toks =
    match src with 
    | [] -> 
        toks @ [EOF]
    | h :: t ->
        match h with
        | "PUSH" -> print_endline "PUSH"; lexe t toks @ [PUSH]
        | "BIND" -> print_endline "BIND"; lexe t toks @ [BIND]
        | _ -> match int_of_string_opt h with
            | Some n -> Printf.printf "NUMBER(%d)\n" n; lexe t toks @ [NUMBER n]
            | None -> Printf.printf "ID(%s)\n" h; lexe t toks @ [ID h]

File: src/vm/src/lexer.go
package main;

import (
	"fmt"
	"strconv"
)

func lexe(vm Vm) []Token {
	tokens :=  make([]Token, 0, len(vm.src));
	
	for _, v := range vm.src {
		switch v {
		case "PUSH":
			fmt.Printf("PUSH\n")
			tokens = append(tokens, PUSH);
		case "BIND":
			fmt.Printf("BIND\n")
			tokens = append(tokens, BIND);
		default:
			n, err := strconv.Atoi(v);
			if err != nil {
				fmt.Printf("ID: %s\n", v)
				tokens = append(tokens, ID);
			} else {
				fmt.Printf("NUMBER: %d\n", n)
				tokens = append(tokens, NUMBER);
			}
		}
	}
	return tokens;
}
